lib/db/feature-tag-store.js

```diff
@@ -0,0 +1,245 @@
+        return rows.map(this.rowToTag);
+    }
+
+    async getTagByTypeAndValue(type, value) {
+        const stopTimer = this.timer('getTagByTypeAndValue');
+        const row = await this.db
+            .first(COLUMNS)
+            .from(TABLE)
+            .where({ type, value })
+            .then(this.rowToTag);
+        stopTimer();
+        return row;
+    }
+
+    async getTag(id) {
+        const stopTimer = this.timer('getTagById');
+        const tag = this.db
+            .first(COLUMNS)
+            .from(TABLE)
+            .where({ id })
+            .then(row => {
+                stopTimer();
+                if (!row) {
+                    throw new NotFoundError(`No tag with id ${id} found`);
+                } else {
+                    return this.rowToTag(row);
+                }
+            });
+        stopTimer();
+        return tag;
+    }
+
+    async hasTag(type, value) {
+        const stopTimer = this.timer('hasTag');
+        const hasTag = this.db
+            .first(COLUMNS)
+            .from(TABLE)
+            .where({ type, value })
+            .then(row => {
+                if (!row) {
+                    stopTimer();
+                    throw new NotFoundError('No tag found');
+                }
+                return {
+                    id: row.id,
+                    value: row.value,
+                    type: row.type,
+                };
+            });
+        stopTimer();
+        return hasTag;
+    }
+
+    async tagFeature(event) {
+        const stopTimer = this.timer('tagFeature');
+        try {
+            await this.hasTag(event.type, event.value);
+        } catch (err) {
+            if (err instanceof NotFoundError) {
+                this.logger.info(
+                    `Tag {type: ${event.type}, value: ${event.value}} did not exist. Creating.`,
+                );
+                await this._createTag(event);
+            } else {
+                this.logger.info('Already existed');
+            }
+        }
+        try {
+            const tag = await this.getTagByTypeAndValue(
+                event.type,
+                event.value,
+            );
+            await this.db(FEATURE_TAG_TABLE)
+                .insert({
+                    feature_name: event.feature_name,
+                    tag_id: tag.id,
+                })
+                .onConflict(['feature_name', 'tag_id'])
+                .ignore();
+            await this.eventStore.store({
+                type: FEATURE_TAGGED,
+                createdBy: event.createdBy || 'unleash-system',
+                data: {
+                    feature_name: event.feature_name,
+                    tag: {
+                        type: event.type,
+                        value: event.value,
+                    },
+                },
+            });
+            stopTimer();
+            return tag;
+        } catch (err) {
+            this.logger.error(err);
+            return {};
```

Comment:
Why does it make sense to return an empty object here?

Reply:
Good question; should probably throw instead; something has definitely gone wrong if we get here

Reply:
I've removed the try/catch here, the calling controller already wrapped this in a try/catch with a call to `handleErrors` in the catch clause, which will cause an error here to yield 500 to the user.